File: ruby/isogram/isogram.rb
# frozen_string_literal: true

module Isogram
  WORD_CHARACTERS = /\w/.freeze
  private_constant :WORD_CHARACTERS

  module_function

  def isogram?(input)
    isogram_letters = input.downcase.scan(WORD_CHARACTERS)
    isogram_letters.uniq == isogram_letters
  end
end

File: javascript/isogram/isogram.js
const WORD_CHARACTERS = /\w/g

export const isIsogram = input => {
  if (input === '') {
    return true
  }

  const isogramLetters = input.toLowerCase().match(WORD_CHARACTERS)
  return isEqual(uniq(isogramLetters), isogramLetters)
};

function uniq(array) {
  return [...new Set(array)]
}

function isEqual(array1, array2) {
  return (
    array1.length === array2.length &&
    array1.every(isContainedIn(array2))
  )
}

function isContainedIn(array) {
  return (char, index) => {
    return array[index] === char
  }
}
